### PR TITLE
Add initial repository support using loopback-datasource-juggler

### PR DESCRIPTION
**Pull request by raymondfeng**
The PR contains a few parts:

1. Define the new interfaces/classes for Repository/DataSource/Connector/Model/Type/Mixin
2. Implement basic types
3. Declare TS classes/interfaces for legacy juggler constructs
4. Implement CRUDRespository based on legacy juggler/connectors
5. Implement `@repository` decorator 

Basic tests are added. The coverage will be improved as follow-up work.

### CODE REVIEW

File: packages/juggler/package.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "@loopback/juggler",
+  "version": "4.0.0-alpha.1",
+  "description": "Juggler for LoopBack.next",
+  "main": "index",
+  "scripts": {
+    "acceptance": "mocha --opts ../../test/mocha.opts 'test/acceptance/**/*.ts'",
+    "build": "npm run build:lib && npm run build:lib6",
+    "build:lib": "tsc --target es2017 --outDir lib",
+    "build:lib6": "tsc --target es2015 --outDir lib6",
```
Comment (bajtos) [inline packages/juggler/package.json]: Please use the new build setup from #300. You will also need to add `packages/juggler/index.ts` file.

File: packages/juggler/package.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "clean": "rm -f loopback-juggler*.tgz && rm -rf lib*",
+    "prepublish": "npm run build",
+    "pretest": "npm run build",
+    "test": "mocha --opts ../../test/mocha.opts 'test/unit/**/*.ts' 'test/acceptance/**/*.ts'",
+    "unit": "mocha --opts ../../test/mocha.opts 'test/unit/**/*.ts'",
+    "verify": "npm run clean && npm pack && tar tf loopback-juggler*.tgz | sort"
```
Comment (bajtos) [inline packages/juggler/package.json]: This is the "verify" script that's used by other packages now:

```json
{
 "verify": "npm pack && tar xf loopback-authentication*.tgz && tree package && npm run clean"
}
```

File: packages/juggler/src/common.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Type alias for Node.js options object
+ */
+export type Options = AnyObject | null | undefined;
+
+/**
+ * Type alias for Node.js callback functions
+ */
+export type Callback<T> =
+  (err: Error | string | null | undefined, result: T) => void;
```
Comment (bajtos) [inline packages/juggler/src/common.ts]: I think `result` should be optional, because it's usually not provided on errors.

```ts
(err: Error | string | null | undefined, result?: T) => void;
```

File: packages/juggler/src/connector.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Common properties/operations for connectors
+ */
+export interface Connector {
+  name: string; // Name/type of the connector
+  configModel?: Model; // The configuration model
+  interfaces?: string[]; // A list of interfaces implemented by the connector
+  connect(): Promise<AnyType>; // Connect to the underlying system
+  disconnect(): Promise<AnyType>; // Disconnect from the underlying system
+  ping(): Promise<AnyType>; // Ping the underlying system
```
Comment (bajtos) [inline packages/juggler/src/connector.ts]: What are these three promises returning? I would expect that at least connect and disconnect are not returning anything, in which case `Promise<void>` is a more accurate description.

File: packages/juggler/src/legacy.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: juggler
```
Comment (bajtos) [inline packages/juggler/src/legacy.ts]: I think this should be `@loopback/juggler`? Please fix the module name in all new files if needed.

Also, can we use a more descriptive file name than `legacy.ts`? For example `legacy-dao-bridge.ts`.

File: packages/juggler/src/loopback-datasource-juggler.ts [hunk 6]
```diff
@@ -0,0 +1,662 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import { Class, Options, Callback, AnyObject, AnyType } from './common';
+
+export declare namespace juggler {
```
Comment (bajtos) [inline packages/juggler/src/loopback-datasource-juggler.ts]: AFAIK, namespaces are not needed when using Node.js `require` system, you can export types directly. 

```js
// loopback-datasource-juggler.ts
export type PromiseOrVoid<T> = Promise<T> | void;

// usage
import * as juggler from './loopback-datasource-juggler';
import {PromiseOrVoid} from './loopback-datasource-juggler';
```

Additionally, can we move this file into loopback-datasource-juggler module, to keep it in sync with the implementation? I am fine to do that later, after this initial pull request has landed.

Thoughts?

File: packages/juggler/src/loopback-datasource-juggler.ts [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+    reload(options?: Options, callback?: Callback<PersistedData>): PromiseOrVoid<PersistedData>;
+
+    /**
+     * Set the correct `id` property for the `PersistedModel`. Uses the `setId` method if the model is attached to
+     * connector that defines it.  Otherwise, uses the default lookup.
+     * Override this method to handle complex IDs.
+     *
+     * @param {*} val The `id` value. Will be converted to the type that the `id` property specifies.
+     */
+    setId(val: AnyType): void;
```
Comment (bajtos) [inline packages/juggler/src/loopback-datasource-juggler.ts]: Considering that `EntityRepository ` is already parameterized by `ID` type, can we make `PersistedModel` parameterized by `ID` type too and use `ID` instead of `AnyType` in all methods?

File: packages/juggler/src/model.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+ */
+
+/**
+ * Property definition for a model
+ */
+export class ModelProperty {
+  name: string;
+  type: string | Function | Object | Type<AnyType>; // For example, 'string', String, or {}
+  json?: PropertyForm;
+  store?: PropertyForm;
```
Comment (bajtos) [inline packages/juggler/src/model.ts]: These two config options `json` and `store` look unfamiliar to me. Are they already recognized and implemented in loopback-datasource-juggler? If not, then I propose to remove them from this patch in order to keep the patch focused on providing the existing juggler API, and to make it easier in the future to understand what code was added in order to support `json` and `store` options.

File: packages/juggler/src/model.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+   */
+  toObject(options?: Options): Object {
+    return {};
+  }
+
+  [prop: string]: AnyType;
+}
+
+export interface Persistable {
+  // isNew: boolean;
```
Comment (bajtos) [inline packages/juggler/src/model.ts]: Why is this commented out? Can we remove it entirely?

File: packages/juggler/src/model.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+export abstract class ValueObject extends Model implements Persistable {
+}
+
+/**
+ * Base class for entities which have unique ids
+ */
+export abstract class Entity extends Model implements Persistable {
+  /**
+   * Get the identity value. If the identity is a composite key, returns
+   * an object.
```
Comment (bajtos) [inline packages/juggler/src/model.ts]: I think this won't work well with non-primitive ID values like MongoDB's `ObjectID`. How can the caller of `getId()` distinguish between the following two cases?
  - `getId()` returned an object with all composite-key values
  - `getId()` returned an ObjectID instance as the single ID used by the model

From my experience so far, composite keys are pretty broken in the current juggler. I am proposing to remove support of composite keys in this initial version of loopback-next, and then later research how to implement composite keys properly. I personally prefer to have a smaller feature set where each feature works really well, rather than a large set of features that may or may not work.


cc @ritch @superkhau

File: packages/juggler/src/types/array.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+// Node module: juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Type, AnyType} from './type';
+
+/**
+ * Array type, such as string[]
+ */
+export class ArrayType<T> implements Type<Array<T>> {
```
Comment (bajtos) [inline packages/juggler/src/types/array.ts]: AFAICT, this class is not covered by unit-tests at all.

I am also wondering, how is this new `Type<T>` system integrated with existing juggler code? Is it a good thing that methods like `isInstance` and `isCoercible` are implemented here in loopback-next and not in the juggler?

File: packages/juggler/test/unit/legacy/legacy.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    ds = new DataSource({
+      name: 'db',
+      connector: 'memory',
+    });
+    expect(ds.settings.name).to.eql('db');
+    expect(ds.settings.connector).to.eql('memory');
+  });
+
+  it('creates models', () => {
+    /* tslint:disable:variable-name */
```
Comment (bajtos) [inline packages/juggler/test/unit/legacy/legacy.ts]: Please use `tslint:disable-next-line:variable-name`. IIUC, your version disables the check for the rest of the file. This applies to most (if not all) other usages of `tslint:disable` in this patch.

File: packages/juggler/tsconfig.json [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+  "compilerOptions": {
+    "outDir": "./lib6"
+  },
+  "include": [
+    "src",
+    "examples"
+  ],
+  "exclude": [
+    "node_modules"
+  ]
```
Comment (bajtos) [inline packages/juggler/tsconfig.json]: "exclude" and "compilerOptions" are no longer needed, see #300.

**Comment by raymondfeng** [comment]
@bajtos PTAL.

File: package.json [hunk 14]
```diff
@@ -10,19 +10,24 @@
...
     "@types/request-promise": "^4.1.33",
     "coveralls": "^2.13.1",
     "lerna": "^2.0.0-rc.4",
     "mocha": "^3.2.0",
     "nyc": "^10.3.2",
     "request": "^2.79.0",
     "request-promise": "^4.1.1",
-    "ts-node": "^2.0.0",
+    "ts-node": "^3.0.4",
+    "tsconfig-paths": "^2.1.1",
```
Comment (bajtos) [inline package.json]: We are no longer using tsconfig-paths, please remove.

File: package.json [hunk 15]
```diff
@@ -10,19 +10,24 @@
...
     "tslint": "^5.1.0",
-    "typescript": "^2.1.5"
+    "typescript": "^2.2.1"
+  },
+  "dependencies": {
```
Comment (bajtos) [inline package.json]: Dependencies must be specified in individual package files, never at the top level. The modules specified in top-level `package.json` won't be installed when users install individual package like `npm install @loopback/juggler`.

File: packages/juggler/.gitignore [hunk 16]
```diff
@@ -0,0 +1,3 @@
+*.tgz
+lib*/
+package
```
Comment (bajtos) [inline packages/juggler/.gitignore]: Do we need a package-specific `.gitignore`? How does this differ from the top-level `.gitignore`?

File: packages/juggler/examples/controllers/CustomerController1.ts [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+import {EntityCrudRepository} from '../../src/repository';
+import {repository} from '../../src/decorator';
+import {Customer} from '../models/customer';
+
+/**
+ * Controller for Customer
+ */
+// @controller
+export class CustomerController1 {
```
Comment (bajtos) [inline packages/juggler/examples/controllers/CustomerController1.ts]: The difference between `CustomerController1`, `CustomerController2` and `CustomerController3` are difficult to tell by looking at the names. Could you please use more descriptive names, e.g. `ControllerWithPropertyDI` and `ControllerWithConstructorDI`?

We also agreed to use kebab case for file names, while these two example files are using PascalCase. Is that intentional?

File: packages/juggler/package.json [hunk 18]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "@loopback/juggler",
```
Comment (bajtos) [inline packages/juggler/package.json]: It makes me wonder whether `juggler` is the best name for this package. Have we considered renaming it, e.g. to `@loopback/persistence`, `@loopback/orm`, `@loopback/data-access`, `@loopback/model` or something else?

File: packages/juggler/src/common.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: @loopback/juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * Common types/interfaces such as Class/Constructor/Options/Callback
```
Comment (bajtos) [inline packages/juggler/src/common.ts]: Nitpick: I am proposing to rename this file to `src/common-types.ts` to make the intent more clear.

File: packages/juggler/src/connector.ts [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
```
Comment (bajtos) [inline packages/juggler/src/connector.ts]: `Promise<AnyType>` means that the method returns a promise that can be resolved with any value. I don't see how is that going to be useful for callers of these three methods (`connect`, `disconnect` and `ping`) when each connector can return whatever it wants.

I am proposing to change return type of these three functions to `Promise<void>`.

File: packages/juggler/src/legacy-juggler-bridge.ts [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+import {Entity} from './model';
+import {Filter, Where} from './query';
+import {EntityCrudRepository} from './repository';
+
+function isPromise<T>(p: AnyType): p is Promise<T> {
+  return p !== null && typeof p === 'object' && typeof p.then === 'function';
+}
+
+function getPromise<T>(p: juggler.PromiseOrVoid<T>): Promise<T> {
```
Comment (bajtos) [inline packages/juggler/src/legacy-juggler-bridge.ts]: The function is called `getPromise` but it either returns the argument value or an error (via a promise rejection). That's not the behaviour I'd expect from a getter method. I am proposing to rename the method to something like `ensurePromise`.

File: packages/juggler/src/legacy-juggler-bridge.ts [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    let id;
+    if (typeof entity.getId === 'function') {
+      id = entity.getId();
+    } else {
+      id = entity[idName];
+    }
+    if (id == null) {
+      return this.create(entity, options);
+    } else {
+      return this.replaceById(id, entity, options).
```
Comment (bajtos) [inline packages/juggler/src/legacy-juggler-bridge.ts]: This implementation does not allow the following usage:

```js
// id "pen1" does not exist yet
const inst = new Product({id: 'pen1', name: 'Pen'});
inst.save();
```

I don't mind must, just pointing out an edge case. If it was an intentional decision then you can ignore my comment. If it was not intentional and you want to land this PR soon, then I am fine with changing this later.

File: packages/juggler/test/unit/legacy/legacy-juggler-bridge.ts [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+      name: 'db',
+      connector: 'memory',
+    });
+    expect(ds.settings.name).to.eql('db');
+    expect(ds.settings.connector).to.eql('memory');
+  });
+
+  it('creates models', () => {
+    /* tslint:disable-next-line:variable-name */
+    const Note = <typeof juggler.PersistedModel>
```
Comment (bajtos) [inline packages/juggler/test/unit/legacy/legacy-juggler-bridge.ts]: Can we provide type definitions for legacy API, so that this type cast is not necessary? Also my impression is that `<TYPE>` casting is considered as a bad practice these days and `as TYPE` should be used instead.

**Comment by raymondfeng** [comment]
@bajtos PTAL.

File: packages/authentication/package.json [hunk 24]
```diff
@@ -19,10 +19,10 @@
   "license": "MIT",
   "dependencies": {
     "@loopback/core": "^4.0.0-alpha.5",
+    "@loopback/context": "4.0.0-alpha.6",
```
Comment (bajtos) [inline packages/authentication/package.json]: Please use carrot operator when specifying dependency versions. That way we don't have to change this line whenever a new context version is released.

```
"@loopback/context": "^4.0.0-alpha.6"
````

File: packages/repository/src/common.ts [hunk 25]
```diff
@@ -0,0 +1,59 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * Common types/interfaces such as Class/Constructor/Options/Callback
```
Comment (bajtos) [inline packages/repository/src/common.ts]: I have proposed to rename this file to `src/common-types.ts` to make the intent more clear. I am not sure if you are intentionally keeping the old name `src/common.ts` or not. I am ok with either.

File: packages/repository/src/connector.ts [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Common properties/operations for connectors
+ */
+export interface Connector {
+  name: string; // Name/type of the connector
+  configModel?: Model; // The configuration model
+  interfaces?: string[]; // A list of interfaces implemented by the connector
+  connect(): Promise<AnyType>; // Connect to the underlying system
+  disconnect(): Promise<AnyType>; // Disconnect from the underlying system
+  ping(): Promise<AnyType>; // Ping the underlying system
```
Comment (bajtos) [inline packages/repository/src/connector.ts]: `Promise<AnyType>` means that the method returns a promise that can be resolved with any value. I don't see how is that going to be useful for callers of these three methods (`connect`, `disconnect` and `ping`) when each connector can return whatever it wants.

I am proposing to change return type of these three functions to `Promise<void>`.

If you disagree then please explain your reasoning so that we can discuss.

File: packages/repository/src/legacy-juggler-bridge.ts [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+  const boundModelClass = class extends modelClass {};
+  boundModelClass.attachTo(ds);
+  return boundModelClass;
+}
+
+import {Entity} from './model';
+import {Filter, Where} from './query';
+import {EntityCrudRepository} from './repository';
+
+function isPromise<T>(p: AnyType): p is Promise<T> {
```
Comment (bajtos) [inline packages/repository/src/legacy-juggler-bridge.ts]: Cross-posting my earlier comment:

The function is called `getPromise` but it either returns the argument value or an error (via a promise rejection). That's not the behaviour I'd expect from a getter method. I am proposing to rename the method to something like `ensurePromise`.

File: packages/repository/src/legacy-juggler-bridge.ts [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+    let id;
+    if (typeof entity.getId === 'function') {
+      id = entity.getId();
+    } else {
+      id = entity[idName];
+    }
+    if (id == null) {
+      return this.create(entity, options);
+    } else {
+      return this.replaceById(id, entity, options).
```
Comment (bajtos) [inline packages/repository/src/legacy-juggler-bridge.ts]: Cross-posting my earlier comment, please let me know your thoughts.

This implementation does not allow the following usage:

```js
// id "pen1" does not exist yet
const inst = new Product({id: 'pen1', name: 'Pen'});
inst.save();
```

I don't mind must, just pointing out an edge case. If it was an intentional decision then you can ignore my comment. If it was not intentional and you want to land this PR soon, then I am fine with changing this later.

File: packages/repository/test/unit/legacy/legacy-juggler-bridge.ts [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+      name: 'db',
+      connector: 'memory',
+    });
+    expect(ds.settings.name).to.eql('db');
+    expect(ds.settings.connector).to.eql('memory');
+  });
+
+  it('creates models', () => {
+    /* tslint:disable-next-line:variable-name */
+    const Note = <typeof juggler.PersistedModel>
```
Comment (bajtos) [inline packages/repository/test/unit/legacy/legacy-juggler-bridge.ts]: Cross-posting my earlier comment:

Can we provide type definitions for legacy API, so that this type cast is not necessary? Also my impression is that `<TYPE>` casting is considered as a bad practice these days and `as TYPE` should be used instead.

**Comment by raymondfeng** [comment]
@bajtos I had to unify the references to `reflect-metadata` so that tests for `@repository` can pass. I'm also working with @deepakrkris to coordinate it with https://github.com/strongloop/loopback-next/pull/297.

File: packages/repository/src/legacy-juggler-bridge.ts (hunk 28 above)
Comment (raymondfeng) [inline packages/repository/src/legacy-juggler-bridge.ts]: With the new `Repository` based CRUD operations, I would like to remove `save` from the model instance.

**Comment by raymondfeng** [comment]
@bajtos PATL

File: package.json [hunk 30]
```diff
@@ -10,9 +10,9 @@
   },
   "license": "MIT",
   "devDependencies": {
-    "@types/mocha": "^2.2.38",
     "@types/node": "^7.0.23",
-    "@types/request": "0.0.40",
+    "@types/mocha": "^2.2.39",
+    "@types/request": "0.0.43",
```
Comment (bajtos) [inline package.json]: `"^0.0.43"` so that we don't have to update this line that often.
